triqs/utility/timestamp.hpp

```diff
@@ -0,0 +1,54 @@
+
+/*******************************************************************************
+ *
+ * TRIQS: a Toolbox for Research in Interacting Quantum Systems
+ *
+ * Copyright (C) 2011 by Hugo U.R. Strand
+ *
+ * TRIQS is free software: you can redistribute it and/or modify it under the
+ * terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * TRIQS is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * TRIQS. If not, see <http://www.gnu.org/licenses/>.
+ *
+ ******************************************************************************/
+#pragma once
+
+#include <iomanip>
+#include <chrono>
+#include <sstream>
+
+namespace triqs {
+  namespace utility {
+
+    std::string inline timestamp() {
+      std::ostringstream s;
+      auto now          = std::chrono::system_clock::now();
+      std::time_t now_c = std::chrono::system_clock::to_time_t(now);
+      s << std::put_time(std::localtime(&now_c), "%H:%M:%S");
+      return s.str();
+    }
+
+    std::string inline hours_minutes_seconds_from_seconds(double seconds) {
+      using namespace std::chrono;
+      time_point<system_clock> tp(milliseconds(long(seconds * 1000)) - hours(1));
```

Comment:
Long story.. Turns out that the `time_point` is initiated with the given time plus one hour. I do not know why but this produces the expected output.

Reply:
Somehow it seems to have swallowed my other comment. 
I think this function is unclear in what it does, using system_clock and localtime
only to convert to number of seconds into hours/minutes/seconds. 
Further it will be problematic if used for runtimes longer than a day.

I suggest to change it to
```c++
std::string hours_minutes_seconds_from_seconds(double sec)
{
    using namespace std::chrono;
    using days = duration<int, std::ratio<86400>>;
    auto s = seconds{int(sec)}; 
    auto d = duration_cast<days>(s);
    auto h = duration_cast<hours>(s -= d);
    auto m = duration_cast<minutes>(s -= h);
    s -= m; 
    std::ostringstream sstream;
    sstream.fill('0');
    sstream << std::setw(2) << d.count() << "d:"
       << std::setw(2) << h.count() << "h:"
       << std::setw(2) << m.count() << "m:"
       << std::setw(2) << s.count() << 's';
    return sstream.str(); 
}
```

Reply:
I am sorry but I disagree. 

The function name and call signature describe exactly what it does.

It solves the problem using the standard library routines for handling time, instead of re-implementing existing functionality. I am strongly against re-implementing time conversion by hand, and against re-implementing the time conversion to string, as you suggest above.

For multiple-day cases the number of hours should simply be more than 24, as one is used to in working with HPC queuing systems. I want the format as HHH:MM:SS since this is the direct input one has to give for wall-time to most HPC schedulers.

In the end I think this is a matter of opinion.

Reply:
> Long story.. Turns out that the time_point is initiated with the given time plus one hour. I do not know why but this produces the expected output.

I tried your solution on wandbox.org and it gives a different result than on my local machine. Subtracting one hour seems to be a hardcoded solution depending on the localtime of your machine.

Reply:
> The function name and call signature describe exactly what it does.

We want both the signature **and** the code to be clear. From what I found that functionality does not exist, so we should be explicit. Your solution tries to use the functionality formatting clocks to, while we should consider std::chrono::duration.
There does not seem to be a formatting tool for that.

> For multiple-day cases the number of hours should simply be more than 24, as one is used to in working with HPC queuing systems. I want the format as HHH:MM:SS since this is the direct input one has to give for wall-time to most HPC schedulers.

I agree that format might be a better choice. However the put_time will just reset to 0 Hours if you go past 24.